source/utils/RandomUtils.cc

```diff
@@ -26,22 +26,16 @@ namespace nexus {
   G4ThreeVector Direction(G4double costheta_min, G4double costheta_max,
                           G4double phi_min, G4double phi_max)
   {
-    while (true) {
-      G4double cosTheta  = 2.*G4UniformRand()-1.;
-      if (cosTheta > costheta_min && cosTheta < costheta_max){
-        G4double sinTheta2 = 1. - cosTheta*cosTheta;
-        if (sinTheta2 < 0.)  sinTheta2 = 0.;
-        G4double sinTheta  = std::sqrt(sinTheta2);
-        while (true) {
-          G4double phi = CLHEP::twopi*G4UniformRand();
-          if (phi > phi_min && phi < phi_max){
-            return G4ThreeVector (sinTheta*std::cos(phi),
-                                  sinTheta*std::sin(phi),
-                                  cosTheta).unit();
-          }
-        }
-      }
-    }
+    G4double cosTheta  = (costheta_max - costheta_min) * G4UniformRand() + costheta_min;
```

Comment:
Good.

However, now that we have to go through this same song-and-dance for `phi`, 5 lines lower down, we should say to ourselves: "Hmm, we've written the same pattern out by hand twice. Is it a useful pattern? Should we abstract it? Are we the first people in the history of programming to have come across the need for this? Is this the last time it will be needed in our project? Does `G4UniformRand` already cater for this option?"

Disappointingly, the authors of `G4UniformRand` appear not to have anticipated this need (I was hoping there were two optional parameters with default values 0 and 1).

I would argue that we should add `uniform_rand_in_range` to our collection of utilities, and hide the details in there.